cli/command/service/ps_test.go

```diff
@@ -0,0 +1,93 @@
+package service
+
+import (
+	"testing"
+
+	"bytes"
+
+	"github.com/docker/cli/cli/internal/test"
+	"github.com/docker/cli/opts"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/context"
+)
+
+type fakeClient struct {
+	client.Client
+	serviceListFunc func(context.Context, types.ServiceListOptions) ([]swarm.Service, error)
+}
+
+func (f *fakeClient) ServiceList(ctx context.Context, options types.ServiceListOptions) ([]swarm.Service, error) {
+	if f.serviceListFunc != nil {
+		return f.serviceListFunc(ctx, options)
+	}
+	return nil, nil
+}
+
+func (f *fakeClient) TaskList(ctx context.Context, options types.TaskListOptions) ([]swarm.Task, error) {
+	return nil, nil
+}
+
+func TestCreateFilter(t *testing.T) {
+	client := &fakeClient{
+		serviceListFunc: func(ctx context.Context, options types.ServiceListOptions) ([]swarm.Service, error) {
+			return []swarm.Service{
+				{ID: "abababab"},
+				{ID: "cccccccc", Spec: swarm.ServiceSpec{Annotations: swarm.Annotations{Name: "foo"}}},
+			}, nil
+		},
+	}
+
+	filter := opts.NewFilterOpt()
+	require.NoError(t, filter.Set("node=somenode"))
+	options := psOptions{
+		services: []string{"abababab", "foo", "notfound"},
+		filter:   filter,
+	}
+
+	actual, notfound, err := createFilter(context.Background(), client, options)
+	require.NoError(t, err)
+	assert.Equal(t, notfound, []string{"no such service: notfound"})
+
+	expected := filters.NewArgs()
+	expected.Add("service", "abababab")
+	expected.Add("service", "cccccccc")
+	expected.Add("node", "somenode")
+	assert.Equal(t, expected, actual)
```

Comment:
Can we also test prefix matches? Specifically:

- An unambiguous ID prefix should match
- An ambiguous ID prefix should lead to an error or warning
- A name prefix should never match.